command/agent_test.go

```diff
@@ -620,3 +620,296 @@ listener "tcp" {
+			}
+
+			// Create a config file
+			config := `
+vault {
+  address = "%s"
+	tls_skip_verify = true
+}
+
+auto_auth {
+    method "approle" {
+        mount_path = "auth/approle"
+        config = {
+            role_id_file_path = "%s"
+            secret_id_file_path = "%s"
+						remove_secret_id_file_after_reading = false
+        }
+    }
+
+    sink "file" {
+        config = {
+            path = "%s"
+        }
+    }
+}
+
+%s
+
+%s
+`
+
+			// conditionally set the exit_after_auth flag
+			exitAfterAuth := ""
+			if tc.exitAfterAuth {
+				exitAfterAuth = "exit_after_auth = true"
+			}
+
+			// flatten the template configs
+			templateConfig := strings.Join(templateConfigStrings, " ")
+
+			config = fmt.Sprintf(config, serverClient.Address(), roleIDPath, secretIDPath, sinkPath, templateConfig, exitAfterAuth)
+			configPath := makeTempFile("config.hcl", config)
+			defer os.Remove(configPath)
+
+			// Start the agent
+			ui, cmd := testAgentCommand(t, logger)
+			cmd.client = serverClient
+			cmd.startedCh = make(chan struct{})
+
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				code := cmd.Run([]string{"-config", configPath})
+				if code != 0 {
+					t.Errorf("non-zero return code when running agent: %d", code)
+					t.Logf("STDOUT from agent:\n%s", ui.OutputWriter.String())
+					t.Logf("STDERR from agent:\n%s", ui.ErrorWriter.String())
+				}
+				wg.Done()
+			}()
+
+			select {
+			case <-cmd.startedCh:
+			case <-time.After(5 * time.Second):
+				t.Errorf("timeout")
+			}
+
+			// We need to sleep to give Agent time to render the templates. Without this
+			// sleep, the test will attempt to read the temp dir before Agent has had time
+			// to render and will likely fail the test
+			time.Sleep(5 * time.Second)
+
+			// if using exit_after_auth, then the command will have returned at the
+			// end and no longer be running. If we are not using exit_after_auth, then
+			// we need to shut down the command
+			if !tc.exitAfterAuth {
+				cmd.ShutdownCh <- struct{}{}
+				wg.Wait()
+			}
+
+			//----------------------------------------------------
+			// Perform the tests
+			//----------------------------------------------------
+
+			files, err := ioutil.ReadDir(tmpDir)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if len(files) != len(templatePaths) {
+				t.Fatalf("expected (%d) templates, got (%d)", len(templatePaths), len(files))
+			}
+		})
+	}
+}
+
+var templateContents = `{{ with secret "secret/myapp"}}
+{
+{{ if .Data.data.username}}"username":"{{ .Data.data.username}}",{{ end }}
+{{ if .Data.data.password }}"password":"{{ .Data.data.password }}",{{ end }}
+{{ if .Data.metadata.version}}"version":"{{ .Data.metadata.version }}"{{ end }}
+}
+{{ end }}`
+
+var templateConfigString = `
+template {
+  source      = "%s"
+  destination = "%s/%s"
+}
+`
+
+// request issues HTTP requests.
+func request(t *testing.T, client *api.Client, req *api.Request, expectedStatusCode int) map[string]interface{} {
```

Comment:
This helper looks like the one from `TestAgent_RequireRequestHeader`, can we change that test to use this as well to clean things up a bit?

Reply:
Sure, cleaned up in cb9558b29